Add a way to run an already distributed function

jcuquemelle opened the pull request:
The goal is to enable users running existing code which is already written with a distribution strategy, without needing to refactor it according to the Experiment formalism, but taking benefit of pex handling, skein cluster configuration and other features of TF_yarn

jcuquemelle committed:
Add a way to run an already distributed function

jcuquemelle committed:
cleaning

jcuquemelle committed:
Add distributed pytorch distributed function example

jcuquemelle committed:
lint

jcuquemelle committed:
lint again

jcuquemelle committed:
fix typing

jcuquemelle committed:
hide custom module in run_on_yarn call

jcuquemelle committed:
mutualize code to choose master

jcuquemelle committed:
Add checkpointing

Removed the PYTORCH_DPP_RANK as it gives the impression that all workers
will checkpoint the model, whereas only the worker 0 does it. This is
usually explicitly exposed in various pytorch documentations

jcuquemelle committed:
remove unused var

jcuquemelle committed:
Remove optionality of optimizer in checkpointing

Was originally made for users who would not want to actually checkpoint,
but only save their model, but this will be better done in a specific
function

nateagr commented on tf_yarn/pytorch/model_ckpt.py:
```diff
@@ -57,10 +55,7 @@ def load_ckpt(
     model_dir: str, model: Union[DDP, torch.nn.Module], optimizer: torch.optim.Optimizer,
     epoch: int, **kwargs: Dict[Any, Any]
-) -> Optional[str]:
-    if int(os.environ[PYTORCH_DPP_RANK]) != 0:
```
This will break our code to train CLIP on yarn. Can we have this feature optional ? Like:
if "PYTORCH_DPP_RANK" in os.environ and int(os.environ[PYTORCH_DPP_RANK]) != 0:
    return None

nateagr commented on tf_yarn/examples/pytorch_distributed_example.py:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    client.run_on_yarn(
+        lambda: train_fcn,
```
We can replace the lambda by train_fcn

jcuquemelle replied:
This was due to the _get_experiment function needing a callable[Callable[Any]], but since a modification needed to solve a pickling issue it is now possible to directly pass the train_fcn

nateagr commented on tf_yarn/examples/pytorch_distributed_example.py:
```diff
@@ -0,0 +1,127 @@
+            output = F.log_softmax(x, dim=1)
+            return output
+
+    # Set of exclusive GPU for the process
```
No support to use several GPUs on a GPU machine with several GPUs ?

jcuquemelle replied:
FAR

nateagr commented on tf_yarn/distributed/task.py:
```diff
@@ -0,0 +1,59 @@
+    addr, port = choose_master(client, task_id)
+    _logger.info(f'master: {addr}:{port}')
+    cluster_tasks = _get_cluster_tasks(client)
+    n_workers_per_executor = 1
```
Why is this hardcoded to 1 ?

jcuquemelle replied:
FAR

jcuquemelle commented on tf_yarn/_task_commons.py:
```diff
@@ -90,3 +94,19 @@ def get_task_description() -> typing.Tuple[str, int]:
     return task_type, task_id
+
+
+def choose_master(client: skein.ApplicationClient, rank: int) -> Tuple[str, int]:
```
_setup_master has been updated to use this method (_setup_master = choose_master + set env variables)

jcuquemelle committed:
add a barrier before destroying process group

Without this, the group was destroyed too soon
and rank0 task would hang waiting to get results
from other tasks

jcuquemelle committed:
Merge branch 'master' into run_already_distributed_code

jcuquemelle committed:
fix pytorch tests and lint

jcuquemelle committed:
Add multiprocessing on same executor for tf-yarn/distributed

jcuquemelle committed:
Add timing to various places during launching job on yarn
